api_mongoDB.go

```diff
@@ -379,12 +381,83 @@ func GetOneCustomDataStructure(collName string, filter bson.M) (bson.M, error) {
 	return result, err
 }
 
-func PutOneCustomDataStructure(collName string, filter bson.M, putData interface{}) bool {
+func PutOneCustomDataStructure(collName string, filter bson.M, putData interface{}) (bool, error) {
 	collection := Client.Database(dbName).Collection(collName)
 
 	var checkItem map[string]interface{}
 	collection.FindOne(context.TODO(), filter).Decode(&checkItem)
 
+	if checkItem == nil {
+		_, err := collection.InsertOne(context.TODO(), putData)
+		if err != nil {
+			logger.MongoDBLog.Println("insert failed : ", err)
+			return false, err
+		}
+		return true, nil
+	} else {
+		collection.UpdateOne(context.TODO(), filter, bson.M{"$set": putData})
+		return true, nil
+	}
+}
+
+func CreateIndex(collName string, keyField string) (bool, error) {
+	collection := Client.Database(dbName).Collection(collName)
+
+	index := mongo.IndexModel{
+		Keys:    bsonx.Doc{{Key: keyField, Value: bsonx.Int32(1)}},
+		Options: options.Index().SetUnique(true),
+	}
+
+	idx, err := collection.Indexes().CreateOne(context.Background(), index)
+	if err != nil {
+		logger.MongoDBLog.Error("Create Index failed : ", keyField, err)
+		return false, err
+	}
+
+	logger.MongoDBLog.Println("index : ", idx)
+
+	return true, nil
+}
+
+func RestfulAPIPatchOneTimeout(collName string, filter bson.M, putData map[string]interface{}, timeout int32, timeField string) bool {
+	collection := Client.Database(dbName).Collection(collName)
+	var checkItem map[string]interface{}
+
+	cursor, err := collection.Indexes().List(context.TODO())
+
+	if err != nil {
+		logger.MongoDBLog.Fatal(err)
+	}
+
+	var result []bson.M
+	if err = cursor.All(context.TODO(), &result); err != nil {
+		logger.MongoDBLog.Fatal(err)
+	}
+
+	for _, v := range result {
+		for k1, v1 := range v {
+			valStr := fmt.Sprint(v1)
+			if (k1 == "name") && strings.Contains(valStr, timeField) {
```

Comment:
can we have k1 field as parameter in the function ?

Reply:
The K1 field will be "name" as will be explained in the comments